### PR TITLE
Expense APIs - Part 2

### PR DESCRIPTION
**Pull request by shalapatil**
## Notion card
https://www.notion.so/saeloun/Backend-Expense-Module-254a3efcf330480cabea0331fed1a20f

## Summary
This PR has all the expenses APIs
Base work required for this PR is done in  https://github.com/saeloun/miru-web/pull/933

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor (non-breaking and retains same functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [ ] I have annotated changes in the PR that are relevant to the reviewer
- [x] I have added automated tests for my code

### CODE REVIEW

**Comment by supriya3105** [comment]
@shalapatil Please add the details to the PR like related notion ticket and description.

File: app/models/vendor.rb [hunk 1]
```diff
@@ -19,6 +19,12 @@
 #  fk_rails_...  (company_id => companies.id)
 #
 class Vendor < ApplicationRecord
+  after_commit :reindex_expenses
+
+  def reindex_expenses
+    expenses.reindex
+  end
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/models/vendor.rb:26]: will add in future a one callbacks method and use that when required. as looks like nothing else will come for now

File: app/services/expenses/fetch_service.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     end
+
+  def process
+    @expenses = search_expenses
+    {
+      expenses:,
+      pagination_details:,
+      vendors: current_company.vendors,
+      categories: ExpenseCategory.defaults.order(:created_at) +
+                    current_company.expense_categories.order(:created_at)
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/services/expenses/fetch_service.rb:20]: even i was thinking of that, but was confused. will do

File: app/views/internal_api/v1/expenses/index.json.jbuilder [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+json.key_format! camelize: :lower
+json.deep_format_keys!
+
+json.expenses expenses do |expense|
+  json.extract! expense, :id, :amount, :expense_type, :category_name, :vendor_name
+  json.date DateTime.parse(expense.date).strftime("%m.%d.%Y")
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/views/internal_api/v1/expenses/index.json.jbuilder]: Done

File: spec/requests/internal_api/v1/expenses/create_spec.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe "InternalApi::V1::Expense#create", type: :request do
+  let(:company) { create(:company) }
+  let(:user) { create(:user, current_workspace_id: company.id) }
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply spec/requests/internal_api/v1/expenses/create_spec.rb]: done

File: app/services/expenses/fetch_service.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  attr_accessor :expenses
+
+  def initialize(current_company, params)
+    @params = params
+    @current_company = current_company
+    @expenses = nil
+     end
+
+  def process
+    @expenses = search_expenses
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/services/expenses/fetch_service.rb]: done

File: app/models/expense_category.rb [hunk 6]
```diff
@@ -20,8 +20,16 @@
 #  fk_rails_...  (company_id => companies.id)
 #
 class ExpenseCategory < ApplicationRecord
+  after_commit :reindex_expenses
+
+  def reindex_expenses
+    expenses.reindex
+  end
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/models/expense_category.rb]: done

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def show
+    authorize expense
+    render :show, locals: {
+      expense: ExpensePresenter.new(expense).snippet
+    }
+  end
+
+  def index
```
[the remark this answers is not in the document]
Comment (shalapatil) [reply app/controllers/internal_api/v1/expenses_controller.rb]: done

**Comment by vipulnsward** [comment]
@shalapatil is this ready for review now after recent merges?

**Comment by shalapatil** [comment]
@vipulnsward Yes, this is ready for review!

File: app/services/expenses/fetch_service.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (vipulnsward) [reply app/services/expenses/fetch_service.rb:13]: I moved it back here. Idea is that initializer can be called once, but everytime we want to fetch records we can call process to fetch these on a fresh basis.

**Comment by shalapatil** [comment]
@vipulnsward ack. Please approve it, so that I can merge it.

**Review by vipulnsward — APPROVED** [review]

**Pull request merged**